docs/typescript/testing.md

```diff
@@ -2,92 +2,246 @@
+<details>
+<summary>
+Workflow implementation
+</summary>
+
+<!--SNIPSTART typescript-timer-reminder-workflow-->
+<!--SNIPEND-->
+
+</details>
+
+```ts
+test('countdownWorkflow sends reminder email if processing does not complete in time', async () => {
+  // NOTE: this tests doesn't actually take days to complete, the test environment starts a test
+  // server that automatically skips time when there are no running activities.
+  let emailSent = false;
+  // createActivities defintion omitted for brevity
+  const activities: ReturnType<typeof createActivities> = {
+    async processOrder() {
+      // Test server switches to "normal" time while an activity is executing.
+      // Call `sleep` to skip time by "2 days".
+      await testEnv.sleep('2 days');
+    },
+    async sendNotificationEmail() {
+      emailSent = true;
+    },
+  };
+  const worker = await Worker.create({
+    connection: testEnv.nativeConnection,
+    taskQueue: 'test',
+    workflowsPath: require.resolve('../workflows'),
+    activities,
   });
+  await worker.runUntil(
+    testEnv.workflowClient.execute(processOrderWorkflow, {
+      workflowId: uuid(),
+      taskQueue: 'test',
+      args: [
+        {
+          orderProcessingMS: ms('3 days'),
+          sendDelayedEmailTimeoutMS: ms('1 day'),
+        },
+      ],
+    })
+  );
+  expect(emailSent).toBe(true);
+});
+```
+
+### Test arbitrary functions in Workflow context
+
+In case you need to test a function in your Workflow code that's not exported in [`workflowsPath`](https://typescript.temporal.io/api/interfaces/worker.workeroptions/#workflowspath), export it in a different path and register it with the Worker.
+
+`workflows/file-with-workflow-function-to-test.ts`
+
+```ts
+import * as wf from '@temporalio/workflow';
+import { someWorkflowToRunAsChild } from './some-workflow';
+
+export { someWorkflowToRunAsChild }; // Must be re-exported here for Worker registration
+
+export async function functionToTest() {
+  await wf.executeChild(someWorkflowToRunAsChild);
+  // Other test code
+}
+```
+
+`test.ts`
+
+```ts
+const worker = await Worker.create({
+  ...someOtherOptions,
+  connection: testEnv.nativeConnection,
+  workflowPath: require.resolve('./workflows/file-workflow-function-to-test'),
 });
+
+await worker.runUntil(
+  testEnv.workflowClient.execute(functionToTest, workflowOptions)
+);
+```
+
+### Asserting from Workflow code
+
+In some cases it's useful to assert directly in Workflow context.
+
+The Workflow context is injected with the Node.js [`assert`](https://nodejs.org/api/assert.html) module and can be imported as any other module.
+
+By default, failed `assert` statement throw `AssertionError`s which cause Workflow Tasks to fail and be indefinitely retried.
+To prevent this, make sure to use `workflowInterceptorModules` from `@temporalio/testing`, these interceptors will catch `AssertionError`s and turn them into `ApplicationFailure`s which fail the entire Workflow Execution.
```

Comment:
```suggestion
By default, failed `assert` statements throw `AssertionError`s which cause Workflow Tasks to fail and be indefinitely retried.
To prevent this, use `workflowInterceptorModules` from `@temporalio/testing`. These interceptors catch `AssertionError`s and turn them into `ApplicationFailure`s that fail the entire Workflow Execution (not just the Workflow Task).
```